Refactor driver/compile_common

gasche opened the pull request:
I was working on a separate PR that would add a new field to the `Compile_common.info` structure, and I realized that there were small aspects of the API that I don't like much. Given that it was only added in the development version, it is time to consider small refactorings before the code reaches the end-users.

I would be interested in particular in the opinion of @Drup and @sliquister (if they have time) who worked on the original code.

This PR is best looked at commit-per-commit, and it may be the case that you like some of them and reject some others (I'm willing to remove some parts). All the commits should be exactly semantics-preserving.

I think that "consistently use `_` as word separator", "rename init_path -> native", "whitespace refactoring" and "store 'native' in 'info'" are clear wins. I believe that removing `ppf_dump` is a nice-to-have (it was a value relying on a lifetime-limited resource with unclear guarantees on liveness), but it's less clear-cut. The `make/init` split is the weakest part of the PR, and I'm interested in opinions on the final refactoring.

gasche committed:
driver: consistently use _ as word separator

gasche committed:
driver/compile_common: rename init_path -> native

The previous label name made little sense to me, and in fact
I thought it meant that the boolean indicates whether the path should
be initialized or not.

gasche committed:
driver/compile_common: whitespace refactoring

gasche committed:
driver/compile_common: store 'native' as part of the 'info' structure

gasche committed:
driver/compile_common: refactor interface/implementation entry points

gasche committed:
Changes mention

Drup commented:
I'm happy to see some iteration on this new API! I had a quick look through the changes, I like all of them except the make/init split. 

I don't mind the idea of separating info-recording and side effects, but I really dislike having the `env` field mutable. I'm also not convinced it's really safer that way. People could just forget to call `init`, and it really should be an `Env.t option`. What's your goal exactly here ? Would it be solved by a `reinit : info -> info` function ?

v-gb commented on driver/compile.ml:
```diff
@@ -23,17 +23,18 @@ let interface = Compile_common.interface ~tool_name
 (** Bytecode compilation backend for .ml files. *)
 
 let to_bytecode i (typedtree, coercion) =
+  Compmisc.with_ppf_dump ~file_prefix:(Compile_common.impl i) @@ fun ppf_dump ->
```
When printing to a file, `Compmisc.with_ppf_dump` truncates said file, so this change looks like it'd break -dump-into-file or whatever the option ended up being called.

v-gb commented on driver/compile_common.mli:
```diff
@@ -28,7 +28,7 @@ type info = {
+  native:bool ->
   tool_name:string ->
   source_file:string -> output_prefix:string -> info
 (** [init ppf ~init_path ~tool_name ~source_file ~output_prefix] initializes
```
The doc string should be updated.

gasche replied:
Thanks for the catch, fixed.

v-gb commented on driver/compile_common.ml:
```diff
@@ -43,6 +43,7 @@ let init ppf_dump ~native ~tool_name ~source_file ~output_prefix =
     source_file;
     ppf_dump;
     tool_name;
+    native;
```
This changes seems a bit weird, in the current state of the code: in `interface`, we always pass `~native:false`. I don't really know why things work this way (perhaps it makes no difference either way for type-checking?), but it seems like we were locally lying before, and now this generalizes the lie.

gasche replied:
With the final state of the PR, `Compile_common.interface` takes an `info` parameter instead of cooking it itself, so this problem does not occur (Compile calls with native=false, Optcompile with native=true), so this problem doesn't exist anymore.

v-gb commented:
The changes you declared as clear wins look good to me.

The ppf change has the problem I indicated inline.

I am not sure about the create/init split, because it makes it unclear who's responsible for initializing the global state. And FWIW, `module_of_filename` is not a pure function (can print warnings), so even at the final state, `make_info` is not a pure function.

gasche commented:
My idea at the beginning was that the make/init split was probably necessary to generalize `info` as an input type to more functions, to be able to innocuously create `info` values and have them initialized deeper in the system. But that doesn't really fly, and I will get rid of the commit, as well as the `ppf_dump` removal for the reason @sliquister pointed out (thanks!).

On the other hand, I think that the `make/init` split is attacking a real problem with the current API, which is the way the creation of an `info` value is tied to a time-limited environment side-effect. (I'm not talking about the formatter's lifetime here, which if broken is easy to detect, but about the environment state.) It is highly unsafe to create several `info` values and use them in parallel, for example; this violates natural assumptions about reification into the `info` value. I don't have the time to do any better than rolling back my splitting commit, but hopefully in 2019 we will remove more ambient state from the compiler implementation.

v-gb commented:
Ok. Interleaved compilations is one problem, but even sequential independent compilations don't work as there's global state that can't be reset without patching the compiler (there's state in asmcomp about linking that's not reset (or there was a year or two ago at least), plus the entirety of command line parsing).

gasche commented:
I made the suggested changes. Thanks!

(I did think about a `mutable : Env.t option` or even a `mutable : Env.t Lazy.t`, but given that we can never get a clean `init/make` anyway I gave up for now.)

@sliquister: I'm thinking of turning the Formatter parameter of `init` into a labelled argument (`~ppf_dump`) for consistency, would that work for you?

gasche committed:
Compile_common.init: make ppf_dump a labelled argument

gasche commented:
I made a last change (last commit), which is to turn `Compile_common.init : ... -> info` into `Compile_common.with_info : ... -> (info -> 'a) -> 'a`. Usually I'm not fond of the more complex types of continuation-passing APIs, but in that case this nicely encapsulates the idea that `info` files, as they are bound to some global state, have a limited validity. In particular, it lets me hide the `ppf_dump` allocation within this call -- expecting an additional `dump_ext` parameter that determines the extension (and thus the name) of the dump file.

gasche committed:
Compile_common: turn `init` into continuation-passing `with_info`

I think this clarifies the fact that the `info` value is a resource that
has a limited lifetime. The new API lets us create (and close) the ppf_dump
directly from `Compile_common.with_info`, simplifying the API for user.

Drup commented:
The issue with `with_info` is that the underlying implementation doesn't match the API. The compiler let lot's of pieces of state and resources around that don't really get reset if you call `with_info` again. 

That being said, that API definitely seems like a good goal, so I'm not really opposed to it. I guess I would prefer if there were explicit comments about the fact that this definitely doesn't encapsulate the complete state of the compiler's pipeline.

gasche commented:
I just added the following to the documentation comment:

```diff
(** [with_info ~native ~tool_name ~source_file ~output_prefix ~dump_ext k]
   invokes its continuation [k] with an [info] structure built from
   its input, after initializing various global variables. This info
   structure and the initialized global state are not valid anymore
   after the continuation returns.

+   Due to current implementation limitations in the compiler, it is
+   unsafe to try to compile several distinct compilation units by
+   calling [with_info] several times.
*)
```

Drup reviewed (approved):
Looks good!
